src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs

```diff
@@ -31,20 +35,20 @@ internal class RecoverableTextAndVersion : ValueSource<TextAndVersion>, ITextVer
             _storageService = storageService;
         }
 
-        private SemaphoreSlim Gate => LazyInitialization.EnsureInitialized(ref _gateDoNotAccessDirectly, SemaphoreSlimFactory.Instance);
+        private SemaphoreSlim Gate => LazyInitialization.EnsureInitialized(ref _lazyGate, SemaphoreSlimFactory.Instance);
 
-        public ITemporaryTextStorage Storage => _text?.Storage;
+        public ITemporaryTextStorage? Storage => _text?.Storage;
 
-        public override bool TryGetValue(out TextAndVersion value)
+        public override bool TryGetValue([MaybeNullWhen(false)]out TextAndVersion value)
```

Comment:
```suggestion
        public override bool TryGetValue([NotNullWhen(true)] out TextAndVersion? value)
```

Reply:
`Dictionary<K,V>` uses `[MaybeNullWhen(false)]`:

https://source.dot.net/#System.Private.CoreLib/Dictionary.cs,903

Reply:
`Dictionary<TKey, TValue>` is required to use `MaybeNullWhen` because there is no source representation for the nullable form of the type. This override uses a concrete type instead, so it can use the source annotation for clarity.

Reply:
I see, so keep MaybeNullWhen on the base but use NotNullWhen on the derived types that substitute the type. Makes sense.

Reply:
Right. It's more obvious on things like `IEquatable<T>.Equals([AllowNull] T)` (the implementation drops `AllowNull` and uses `?` instead). This is the equivalent situation for an `out` parameter.

Reply:
Actually, shouldn't NotNullWhen be used only with nullable types? That is 

`TryGetValue([NotNullWhen(true)]out TextAndVersion? value)`

rather than

`TryGetValue([NotNullWhen(true)]out TextAndVersion value)`?

Because if the parameter type is not nullable then saying that it's not-null when the method returns true is redundant.

Reply:
But then I can;'t change nullability of the parameter in the overload (vs the base).

Reply:
> Actually, shouldn't NotNullWhen be used only with nullable types? 

Yes, my original suggestion changed the attribute and also changed the parameter type.

Reply:
Changing the nullability of out parameter is currently not allowed by the compiler, so I'll leave MaybeNullWhen